sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2457,164 +2451,133 @@ class Analyzer(override val catalogManager: CatalogManager)
+        resolveOperatorWithAggregate(maybeResolved, agg, (newExprs, newChild) => {
+          val newSortOrder = sortOrder.zip(newExprs).map {
+            case (sortOrder, expr) => sortOrder.copy(child = expr)
           }
+          Sort(newSortOrder, global, newChild)
+        })
+    }
 
-          val sortOrdersMap = unresolvedSortOrders
-            .map(new TreeNodeRef(_))
-            .zip(evaluatedOrderings)
-            .toMap
-          val finalSortOrders = sortOrder.map(s => sortOrdersMap.getOrElse(new TreeNodeRef(s), s))
+    /**
+     * Resolves the given expressions as if they are in the given Aggregate operator, which means
+     * the column can be resolved using `agg.child` and aggregate functions/grouping columns are
+     * allowed. It returns a list of named expressions that need to be appended to
+     * `agg.aggregateExpressions`, and the list of resolved expressions.
+     */
+    def resolveExprsWithAggregate(
+        exprs: Seq[Expression],
+        agg: Aggregate): (Seq[NamedExpression], Seq[Expression]) = {
+      def resolveCol(input: Expression): Expression = {
+        input.transform {
+          case u: UnresolvedAttribute =>
+            try {
+              // Resolve the column and wrap it with `TempResolvedColumn`. If the resolved column
+              // doesn't end up with as aggregate function input or grouping column, we should
+              // undo the column resolution to avoid confusing error message. For example, if
+              // a table `t` has two columns `c1` and `c2`, for query `SELECT ... FROM t
+              // GROUP BY c1 HAVING c2 = 0`, even though we can resolve column `c2` here, we
+              // should undo it later and fail with "Column c2 not found".
+              agg.child.resolve(u.nameParts, resolver).map(TempResolvedColumn(_, u.nameParts))
+                .getOrElse(u)
+            } catch {
+              case _: AnalysisException => u
+            }
+        }
+      }
 
-          // Since we don't rely on sort.resolved as the stop condition for this rule,
-          // we need to check this and prevent applying this rule multiple times
-          if (sortOrder == finalSortOrders) {
-            sort
-          } else {
-            Project(aggregate.output,
-              Sort(finalSortOrders, global,
-                aggregate.copy(aggregateExpressions = originalAggExprs ++ needsPushDown)))
-          }
-        } catch {
-          // Attempting to resolve in the aggregate can result in ambiguity.  When this happens,
-          // just return the original plan.
-          case ae: AnalysisException => sort
+      def resolveSubQuery(input: Expression): Expression = {
+        if (SubqueryExpression.hasSubquery(input)) {
+          val fake = Project(Alias(input, "fake")() :: Nil, agg.child)
+          ResolveSubquery(fake).asInstanceOf[Project].projectList.head.asInstanceOf[Alias].child
```

Comment:
The children of the resolved subquery that are not aggregate expressions should also be wrapped in `TempResolvedColumn` (but it seems the current logic can't handle this case with proper error messages either. This can be a follow-up in the future)
```sql
select c1 from t1 group by c1 having (select sum(c2) from t2 where t1.c2 = t2.c1) > 0
-- org.apache.spark.sql.AnalysisException: Resolved attribute(s) c2#10 missing from c1#9 in operator
-- !Filter (scalar-subquery#8 [c2#10] > cast(0 as bigint)).;
```